e2e/util/recipe.go

```diff
@@ -0,0 +1,286 @@
+						Values:   []string{"busybox"},
+					},
+				},
+			},
+		},
+	}
+}
+
+// nolint:mnd
+func prepareHooks(namespace string) []*recipe.Hook {
+	return []*recipe.Hook{
+		{
+			Name:           "check-hook",
+			Type:           "check",
+			Namespace:      namespace,
+			NameSelector:   "busybox",
+			SelectResource: "deployment",
+			Timeout:        300,
+			Chks: []*recipe.Check{
+				{
+					Name:      "check-replicas",
+					Condition: "{$.spec.replicas} == {$.status.readyReplicas}",
+				},
+			},
+		},
+		{
+			Name:           "exec-hook",
+			Type:           "exec",
+			Namespace:      namespace,
+			NameSelector:   "busybox",
+			SelectResource: "deployment",
+			Timeout:        300,
+			Ops: []*recipe.Operation{
+				{
+					Name:    "ls",
+					Command: "/bin/sh -c ls",
+				},
+			},
+		},
+	}
+}
+
+func prepareWorkflow(name string) *recipe.Workflow {
+	return &recipe.Workflow{
+		Name: name,
+		Sequence: []map[string]string{
+			{
+				"group": "rg1",
+			},
+		},
+	}
+}
+
+func prepareWorkflows(recipeSpec *config.Recipe) []*recipe.Workflow {
+	backup := prepareWorkflow("backup")
+	restore := prepareWorkflow("restore")
+
+	checkHook := map[string]string{
+		"hook": "check-hook/check-replicas",
+	}
+
+	execHook := map[string]string{
+		"hook": "exec-hook/ls",
+	}
+
+	group := map[string]string{
+		"group": "rg1",
+	}
+
+	if recipeSpec.ExecHook && recipeSpec.CheckHook {
+		addCheckAndExecHooks(backup, checkHook, execHook, group)
+		addCheckAndExecHooks(restore, checkHook, execHook, group)
+
+		return []*recipe.Workflow{
+			backup,
+			restore,
+		}
+	}
+
+	if recipeSpec.CheckHook {
+		addCheckHooks(backup, checkHook, group)
+		addCheckHooks(restore, checkHook, group)
+
+		return []*recipe.Workflow{
+			backup,
+			restore,
+		}
+	}
+
+	if recipeSpec.ExecHook {
+		addExecHooks(backup, execHook, group)
+		addExecHooks(restore, execHook, group)
+	}
+
+	return []*recipe.Workflow{
+		backup,
+		restore,
+	}
+}
+
+// nolint:mnd
+func addCheckHooks(workflow *recipe.Workflow, checkHook, group map[string]string) {
+	seq := make([]map[string]string, 2)
+
+	if workflow.Name == "backup" {
+		seq[0] = checkHook
+
+		seq[1] = group
+	}
+
+	if workflow.Name == "restore" {
+		seq[0] = group
+
+		seq[1] = checkHook
+	}
+
+	workflow.Sequence = seq
+}
+
+// nolint:mnd
+func addExecHooks(workflow *recipe.Workflow, execHook, group map[string]string) {
+	seq := make([]map[string]string, 2)
+
+	if workflow.Name == "backup" {
+		seq[0] = execHook
+		seq[1] = group
+	}
+
+	if workflow.Name == "restore" {
+		seq[0] = group
+		seq[1] = execHook
+	}
+
+	workflow.Sequence = seq
+}
+
+// nolint:mnd
+func addCheckAndExecHooks(workflow *recipe.Workflow, checkHook, execHook, group map[string]string) {
+	seq := make([]map[string]string, 3)
+
+	if workflow.Name == "backup" {
+		seq[0] = checkHook
+		seq[1] = execHook
+		seq[2] = group
+	}
+
+	if workflow.Name == "restore" {
+		seq[0] = group
+		seq[1] = checkHook
+		seq[2] = execHook
+	}
+
+	workflow.Sequence = seq
```

Comment:
So checkHooks and execHooks must be ordered in certain way for backup and restore, which probably makes it harder to create a generic addHook.

But since checkHook and execHook and group are the same type, we can create a generic function accepting one or more hooks and group (map[string]string]), or just set it in the caller directly:

    workflow.Sequence = []map[string]string{checkHook, execHook, group}

It seems that the helpers to add the hooks are not helping much and it will be more clear to do everything in prepareWorkflows, or if you want to help ensure the right order of the group and hooks, add a helper for adding stuff to backup and another helper to adding stuff to restore.